### PR TITLE
Add fuzzy matching support in room searches

### PR DESCRIPTION
**Pull request by gsouquet**
Fixes vector-im/element-web#14410

This adds support for fuzzy matching in room searches

This is a rather simplistic approach to it. There might be some UX changes to this, I have open a discussion on the #element-dev:matrix.org channel

### CODE REVIEW

File: src/stores/room-list/filters/NameFilterCondition.ts [hunk 1]
```diff
@@ -19,16 +19,23 @@ import { FILTER_CHANGED, FilterKind, IFilterCondition } from "./IFilterCondition
...
 /**
  * A filter condition for the room list which reveals rooms of a particular
  * name, or associated name (like a room alias).
  */
 export class NameFilterCondition extends EventEmitter implements IFilterCondition {
     private _search = "";
+    private fuzzy = false;
 
-    constructor() {
+    constructor(options: IOpts = {}) {
```
Comment (turt2live) [inline src/stores/room-list/filters/NameFilterCondition.ts:36]: not sure how we feel about this more generally, but making this `private fuzzy = false` is slightly cleaner and works well with IDEs. The options approach does allow for expandability, however this is internal API and can be changed with relative ease.

File: src/utils/strings.ts [hunk 2]
```diff
@@ -84,3 +86,24 @@ const collator = new Intl.Collator();
 export function compare(a: string, b: string): number {
     return collator.compare(a, b);
 }
+
+/**
+ * A cache for regular expression used in fuzzy matching
+ * @param str The string to inspect
+ * @return {RegExp} a regular expression to inspect `str`
```
Comment (turt2live) [inline src/utils/strings.ts:93]: some words to explain that this is a performance benefit would be appreciated. Also fullstops :)

File: test/utils/strings-test.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { fuzzyMatch } from "../../src/utils/strings";
+
+describe("strings", () => {
+    describe("fuzzyMatch()", () => {
```
Comment (turt2live) [inline test/utils/strings-test.ts:20]: a test that deals with numbers and symbols interspersed with characters would be nice as well, to ensure the regex cleaning works (of which there doesn't appear to be any?) and that it handles non-ascii well (including emoji).

File: src/utils/strings.ts [hunk 4]
```diff
@@ -84,3 +86,24 @@ const collator = new Intl.Collator();
...
+ */
+const fuzzyCache = memoize((str: string): RegExp => {
+    return new RegExp("^"+str.replace(/./g, function(x) {
+        // eslint-disable-next-line
```
Comment (turt2live) [inline src/utils/strings.ts:97]: disable *what* on the next line?

**Review by turt2live — CHANGES_REQUESTED** [review]

File: src/stores/room-list/filters/NameFilterCondition.ts (hunk 1 above)
Comment (gsouquet) [reply src/stores/room-list/filters/NameFilterCondition.ts:36]: I'm not sure if I understand the change that you're suggesting.

As far as I can tell I have created `private fuzzy = false` and this options parameter is only here to let a developer choose the type of matching the `NameFilterCondition` should apply

Could you show me an example of what you had in mind instead?

File: test/utils/strings-test.ts (hunk 3 above)
Comment (gsouquet) [reply test/utils/strings-test.ts:20]: Makes sense, I have added more test cases to deal with more exotic character set.

I'm not sure if I know what regex cleaning is and what you expect it to do in this scenario?

**Comment by novocaine** [comment]
@turt2live bump, can you help clarify some of your comments above please

File: src/stores/room-list/filters/NameFilterCondition.ts (hunk 1 above)
Comment (turt2live) [reply src/stores/room-list/filters/NameFilterCondition.ts:36]: In this case, merging the field declaration with the constructor declaration: `constructor(private fuzzy = false) {`

It's a little bit cleaner for the API contract this part of the code is after.

File: test/utils/strings-test.ts (hunk 3 above)
Comment (turt2live) [reply test/utils/strings-test.ts:19]: "regex cleaning" is my horrible way of referring to the action the regex performs. There should be tests to make sure that the regex does the right thing, where the "right thing" is cleaning/sanitizing the input from what I can see.

**Comment by turt2live** [comment]
Sorry, I had this one slip through the notifications. A mention or review request will always cut through the sea of notifications I deal with.

**Comment by novocaine** [comment]
@gsouquet can you please update this ticket to describe the changes you've made so that we can review them from a product and design perspective?

**Comment by gsouquet** [comment]
There are no changes done to the UI at the moment.

What was originally discussed with Amsha is that fuzzy matching is going to yield more results. So we need to agree on a visual representation of what has "matched", and how to sort those results.

One caveat with the sorting is that the room list tags (categories) already provide an option to update the sorting of tiles (either alphabetically or by last activity)

Screen Shot 2021-10-01 at 15 06 33

Filtering with fuzzy matching would potentially override that sorting to bump the more relevant tiles to the top, which could be quite jarring. But that is definitely something to explore and see what would make the most sense with our current design

**Comment by amshakal** [comment]
Yeah, I think it's worth considering to show search results by relevance alone rather than based on activity or A-Z, given that that's how most search works on different platforms.

**Comment by gsouquet** [comment]
This pull request is now stale and should handle the new search experience. We're also not doing this as part of the work to get the new search experience out of beta

**Pull request closed without merging**